src/zenml/logging/step_logging.py

```diff
@@ -161,6 +330,74 @@ def save_to_file(self) -> None:
                 self.last_save_time = time.time()
 
                 self.disabled = False
+        # merge created files on a given interval (defaults to 10 minutes)
+        # only runs on Immutable Filesystems
+        if (
+            self.artifact_store.config.IS_IMMUTABLE_FILESYSTEM
+            and time.time() - self.last_merge_time > self.merge_files_interval
+        ):
+            try:
+                self.merge_log_files(
+                    self.next_merged_file_name, self.log_files_not_merged
+                )
+            except (OSError, IOError) as e:
+                logger.error(f"Error while trying to roll up logs: {e}")
+            else:
+                self.log_files_not_merged = []
+            finally:
+                self.last_merge_time = time.time()
+
+    def merge_log_files(
+        self,
+        file_name: Optional[str] = None,
+        files: Optional[List[str]] = None,
+    ) -> None:
+        """Merges all log files into one in the given URI.
+
+        Called on the logging context exit.
+
+        Args:
+            file_name: The name of the merged log file.
+            files: The list of log files to merge.
+        """
+        if self.artifact_store.config.IS_IMMUTABLE_FILESYSTEM:
+            files_ = files or self.artifact_store.listdir(self.logs_uri)
+            file_name_ = file_name or f"full_log{LOGS_EXTENSION}"
+            if len(files_) > 1:
+                files_.sort()
+                logger.debug("Log files count: %s", len(files_))
+
+                with TemporaryDirectory() as temp_dir:
+                    try:
+                        local_log_file = os.path.join(temp_dir, file_name_)
+                        # dump all logs to a local file first
+                        with open(local_log_file, "w") as merged_file:
+                            for file in files_:
+                                merged_file.write(
+                                    str(
+                                        _load_file_from_artifact_store(
+                                            os.path.join(
+                                                self.logs_uri, str(file)
+                                            ),
+                                            artifact_store=self.artifact_store,
+                                            mode="r",
+                                        )
+                                    )
+                                )
+
+                        # copy it over to the artifact store
+                        fileio.copy(
+                            local_log_file,
+                            os.path.join(self.logs_uri, file_name_),
+                        )
```

Comment:
Minor suggestion: why not avoid having to store anything in the local filesystem by writing directly into the artifact store?

```suggestion
                        with self.artifact_store.open(os.path.join(self.logs_uri, file_name_), "w") as merged_file:
                            for file in files_:
                                merged_file.write(
                                    str(
                                        _load_file_from_artifact_store(
                                            os.path.join(
                                                self.logs_uri, str(file)
                                            ),
                                            artifact_store=self.artifact_store,
                                            mode="r",
                                        )
                                    )
                                )
```

Reply:
Agree